scio-parquet/src/main/scala/com/spotify/scio/parquet/tensorflow/ParquetExampleIO.scala

```diff
@@ -0,0 +1,147 @@
+  DynamicFileDestinations,
+  FileBasedSink,
+  FileSystems,
+  WriteFiles
+}
+import org.apache.beam.sdk.io.hadoop.format.HadoopFormatIO
+import org.apache.beam.sdk.options.ValueProvider.StaticValueProvider
+import org.apache.beam.sdk.transforms.SimpleFunction
+import org.apache.hadoop.mapreduce.Job
+import org.apache.parquet.filter2.predicate.FilterPredicate
+import org.apache.parquet.hadoop.ParquetInputFormat
+import org.apache.parquet.hadoop.metadata.CompressionCodecName
+import org.tensorflow.example.Example
+
+import scala.collection.JavaConverters._
+
+final case class ParquetExampleIO(path: String) extends ScioIO[Example] {
+  override type ReadP = ParquetExampleIO.ReadParam
+  override type WriteP = ParquetExampleIO.WriteParam
+  override val tapT = TapOf[Example]
+
+  override protected def read(sc: ScioContext, params: ReadP): SCollection[Example] = {
+    val job = Job.getInstance()
+    GcsConnectorUtil.setInputPaths(sc, job, path)
+    job.setInputFormatClass(classOf[ExampleParquetInputFormat])
+    job.getConfiguration.setClass("key.class", classOf[Void], classOf[Void])
+    job.getConfiguration.setClass("value.class", classOf[Example], classOf[Example])
+
+    ParquetInputFormat.setReadSupportClass(job, classOf[ExampleReadSupport])
+    if (params.projection != null) {
+      ExampleParquetInputFormat.setFields(job, params.projection.asJava)
+    }
+    if (params.predicate != null) {
+      ParquetInputFormat.setFilterPredicate(job.getConfiguration, params.predicate)
+    }
+
+    val source = HadoopFormatIO
+      .read[JBoolean, Example]()
+      .withKeyTranslation(new SimpleFunction[Void, JBoolean]() {
+        override def apply(input: Void): JBoolean = true // workaround for NPE
+      })
+      .withConfiguration(job.getConfiguration)
+    sc.wrap(sc.applyInternal(source)).map(_.getValue)
+  }
+
+  override protected def write(data: SCollection[Example], params: WriteP): Tap[Example] = {
+    val job = Job.getInstance()
+    if (ScioUtil.isLocalRunner(data.context.options.getRunner)) {
+      GcsConnectorUtil.setCredentials(job)
+    }
+
+    val resource =
+      FileBasedSink.convertToFileResourceIfPossible(ScioUtil.pathWithShards(path))
+    val prefix = StaticValueProvider.of(resource)
+    val usedFilenamePolicy =
+      DefaultFilenamePolicy.fromStandardParameters(prefix, null, params.suffix, false)
+    val destinations = DynamicFileDestinations.constant[Example](usedFilenamePolicy)
+    val sink = new ParquetExampleSink(
+      prefix,
+      destinations,
+      params.schema,
+      job.getConfiguration,
+      params.compression
+    )
+    val t = WriteFiles.to(sink).withNumShards(params.numShards)
+    data.applyInternal(t)
+    tap(ParquetExampleIO.ReadParam())
+  }
+
+  override def tap(params: ReadP): Tap[Example] =
+    ParquetExapmleTap(ScioUtil.addPartSuffix(path), params)
+}
+
+object ParquetExampleIO {
+  final case class ReadParam private (
+    projection: Seq[String] = null,
```

Comment:
I know we use this `null` thing a bit all over the place ... but I'm always sad to see it 😢